### PR TITLE
feat/staking token sweep

### PR DESCRIPTION
**Pull request by aodhgan**
### Purpose for this PR
as per https://github.com/windranger-io/windranger-treasury/issues/298

### CODE REVIEW

**Comment by CjHare** [comment]
You have unhappy tests 😞

File: contracts/sweep/SweepERC20.sol [hunk 1]
```diff
@@ -32,6 +32,6 @@ abstract contract SweepERC20 is TokenSweep {
             tokenSweepBeneficiary(),
             amount
         );
-        require(result, "SweepERC20: transfer failed");
+        require(result, "SweepERC20: transfer failed"); //
```
Comment (CjHare) [inline contracts/sweep/SweepERC20.sol:35]: question: why the `//`?

File: test/contracts/staking/staking-factory.test.ts [hunk 2]
```diff
@@ -42,26 +62,187 @@ export type StakingPoolLibData = {
     maxTotalPoolStake: number
 }
 
-describe('Staking Pool Factory', () => {
+describe.only('Staking Pool Factory', () => {
```
Comment (CjHare) [inline test/contracts/staking/staking-factory.test.ts:65]: issue: `describe` not `describe.only`

_An extra dot only_

File: test/contracts/staking/staking-factory.test.ts [hunk 3]
```diff
@@ -42,26 +62,187 @@ export type StakingPoolLibData = {
...
+                ).to.be.revertedWith('Pausable: paused')
+            })
+        })
+        after(async () => {
+            await stakingPoolFactory.unpause()
+        })
     })
 
     describe('create pools', () => {
+        before(async () => {
```
Comment (CjHare) [inline test/contracts/staking/staking-factory.test.ts:240]: question: is the `epochStartTimestamp` needed in this changeset or a general fix?

File: contracts/sweep/SweepERC20.sol (hunk 1 above)
Comment (aodhgan) [reply contracts/sweep/SweepERC20.sol:35]: oh was leaving it in here as a reminder to myself -should we be using Safe Transfer here?

File: test/contracts/staking/staking-factory.test.ts (hunk 2 above)
Comment (aodhgan) [reply test/contracts/staking/staking-factory.test.ts:65]: yeah this is a hack i use for dev to run tests faster. this PR should have been marked WIP, apologies if wasted cycles!

File: test/contracts/staking/staking-factory.test.ts (hunk 3 above)
Comment (aodhgan) [reply test/contracts/staking/staking-factory.test.ts:240]: Actually increasing the START_DELAY var negates the need to do this. I guess each hardhat transaction rolls the chain timestamp forward and by the time these tests were being run, the initial timestamp had be surpassed (and the contracts disallow historic timestamps for the creation of pools).

File: contracts/staking/StakingPool.sol [hunk 4]
```diff
@@ -228,14 +230,24 @@ contract StakingPool is
         _withdrawStake();
     }
 
+    function sweepERC20Tokens(address tokens, uint256 amount)
+        external
+        whenNotPaused
+        onlySuperUserRole
```
Comment (aodhgan) [inline contracts/staking/StakingPool.sol:236]: this permissioning will change to onlyOwner

File: contracts/sweep/SweepERC20.sol (hunk 1 above)
Comment (CjHare) [reply contracts/sweep/SweepERC20.sol:35]: Hmmm.... `safeTransfer` 🤔 

We could, but we should apply it consistently i.e. use it everywhere.

Unless you've already started using it (and already imported the lib), I'd raise an issue to apply everywhere once the contracts have been split and downsized.

File: contracts/staking/StakingPool.sol (hunk 4 above)
Comment (CjHare) [reply contracts/staking/StakingPool.sol:236]: 👍

File: test/contracts/staking/staking-factory.test.ts [hunk 5]
```diff
@@ -8,9 +8,16 @@ import {before} from 'mocha'
 import {solidity} from 'ethereum-waffle'
 import {
     StakingPoolFactory,
-    ERC20PresetMinterPauser
+    ERC20PresetMinterPauser,
+    BondMediator,
```
Comment (CjHare) [inline test/contracts/staking/staking-factory.test.ts:12]: issue: unused imports

The linter is picking unused import, remove them please.

**Review by CjHare — COMMENTED** [review]
Looking good, with the permission, do you want to merge this in now & update the access control, or await the other PR and update this one?

File: contracts/sweep/SweepERC20.sol (hunk 1 above)
Comment (aodhgan) [reply contracts/sweep/SweepERC20.sol:35]: opened https://github.com/windranger-io/windranger-treasury/issues/328

File: contracts/staking/StakingPool.sol [hunk 6]
```diff
@@ -298,6 +310,14 @@ contract StakingPool is
         _setRewardsAvailableTimestamp(timestamp);
     }
 
+    function updateTokenSweepBeneficiary(address newBeneficiary)
+        external
+        whenNotPaused
+        onlySuperUserRole
```
Comment (aodhgan) [inline contracts/staking/StakingPool.sol:316]: will change to onlyOwner after #308

**Review by CjHare — APPROVED** [review]
LGTM

**Pull request merged**